### ITCS-5111 Intro to Natural Language Processing

<a href="#Question 1 Answer English" style="color:blue;">Click here to have a look at Answer to Question-1 English</a>

<a href="#Question 1 Answer Italian" style="color:blue;">Click here to have a look at Answer to Question-1 Italian</a>

<a href="#Question 2 Answer-Ngram-model" style="color:blue;">Click here to have a look at Question-2 N-gram model function</a>

<a href="#Question 2 Answer-Prob" style="color:blue;">Click here to have a look at Question-2 Probability function</a>

<a href="#Question 2 Answer-model-implemetation" style="color:blue;">Click here to have a look at Question-2 N-gram model Implementation function</a>

<a href="#Question 2 Answer-accuracy" style="color:blue;">Click here to have a look at Question-2 Unigram model Accuracy Analysis</a>

<a href="#Question 3" style="color:blue;">Click here to have a look at the Answer for Question-3 Bigram model Accuracy Analysis</a>

<a href="#Question 4" style="color:blue;">Click here to have a look at the Answer for Question-4 Trigram model Accuracy Analysis</a>

In [1]:
import pandas as pd
import numpy as np
import nltk

In [2]:
# Question-1
# Load the course data files for English into a dataframe using Pandas
dataframe_english = pd.read_csv(r"C:\Users\muthu\Desktop\UNCC\Semester-3\NLP\Assignments\Data_Input\CONcreTEXT_trial_EN.tsv", sep = "\t")
dataframe_english

TARGET POS  INDEX                                               TEXT  \
0   achievement   N      3  Bring up academic achievements , awards , and ...   
1   achievement   N      9  Please list people you have helped , your pers...   
2      activate   V      1     Add activated carbon straight to your vodka .    
3      activate   V     15  Place sensors around your garden , and when a ...   
4     adventure   N      9  Look for a partner that shares your level of a...   
..          ...  ..    ...                                                ...   
95        water   N      5  Rinse your face with warm water and pat it dry .    
96          win   V      4  Staying mentally strong means winning half the...   
97          win   V      7  The person who has the highest score wins the ...   
98        woman   N      7  For the most part , men and women wear the sam...   
99        woman   N      3  Look at the woman whom you are listening to fo...   

    MEAN  
0   3.06  
1   3.03  
2   3.83  
3   5.51  
4   2.03  
..   ...  
95  6.91  
96  2.34  
97  4.60  
98  6.29  
99  6.57  

[100 rows x 5 columns]

In [3]:
# Question-1
# Load the course data files for Italian into a dataframe using Pandas
dataframe_italian = pd.read_csv(r"C:\Users\muthu\Desktop\UNCC\Semester-3\NLP\Assignments\Data_Input\CONcreTEXT_trial_IT.tsv", sep = "\t")
dataframe_italian

TARGET POS  INDEX                                               TEXT  \
0   affascinare   V      8  Guardati i piedi o fai finta di essere affasci...   
1   affascinare   V      1  Sei affascinato dal funzionamento della mente ...   
2         amore   N      5  Pensa ai tuoi sentimenti di amore , passione e...   
3         amore   N      7  Quasi tutti sono in grado di ricevere amore e ...   
4          aria   N      4  Accendi il condizionatore sull’ aria fredda ma...   
..          ...  ..    ...                                                ...   
95       verità   N      8  In un modo o nell' altro , la verità viene sem...   
96      viaggio   N      2  Organizza dei viaggi nel fine settimana quando...   
97      viaggio   N      6  Pesa le tue valigie prima del viaggio per evit...   
98        vista   N      6  è molto importante non perdere di vista la pro...   
99        vista   N      9  i conigli hanno un ottimo udito e un' ottima v...   

    MEAN  
0   2.39  
1   2.03  
2   2.71  
3   2.16  
4   5.68  
..   ...  
95  2.53  
96  5.03  
97  4.84  
98  2.22  
99  5.13  

[100 rows x 5 columns]

In [4]:
# Removing Punctuations
from nltk.tokenize import RegexpTokenizer
tokenizer_pattern = RegexpTokenizer('\w+')
tokenizer_pattern

RegexpTokenizer(pattern='\\w+', gaps=False, discard_empty=True, flags=<RegexFlag.UNICODE|DOTALL|MULTILINE: 56>)

In [5]:
# For English Dataframe
# Using NLTK libraries, strip the text of punctuation and lowercase the words, and obtain from the original text a list of words
dataframe_english['TEXT'] = dataframe_english['TEXT'].apply(lambda x: tokenizer_pattern.tokenize(x.lower()))
dataframe_english['TEXT'].head(15)

0     [bring, up, academic, achievements, awards, an...
1     [please, list, people, you, have, helped, your...
2     [add, activated, carbon, straight, to, your, v...
3     [place, sensors, around, your, garden, and, wh...
4     [look, for, a, partner, that, shares, your, le...
5     [most, animated, action, adventure, thriller, ...
6     [if, people, think, you, re, phony, or, puttin...
7     [place, your, tongue, behind, the, hole, in, y...
8     [bake, the, fries, until, they, are, crispy, a...
9     [pour, into, a, soufflé, dish, and, bake, unti...
10    [meditation, can, help, your, mind, and, body,...
11    [they, are, easier, for, the, human, body, to,...
12    [for, example, you, could, join, a, book, club...
13    [if, you, publish, a, book, of, poetry, or, a,...
14    [expenses, include, rent, monthly, loans, car,...
Name: TEXT, dtype: object

In [6]:
# For Italian Dataframe
# Using NLTK libraries, strip the text of punctuation and lowercase the words, and obtain from the original text a list of words
dataframe_italian['TEXT'] = dataframe_italian['TEXT'].apply(lambda x: tokenizer_pattern.tokenize(x.lower()))
dataframe_italian['TEXT'].head(100)

0     [guardati, i, piedi, o, fai, finta, di, essere...
1     [sei, affascinato, dal, funzionamento, della, ...
2     [pensa, ai, tuoi, sentimenti, di, amore, passi...
3     [quasi, tutti, sono, in, grado, di, ricevere, ...
4     [accendi, il, condizionatore, sull, aria, fred...
                            ...                        
95    [in, un, modo, o, nell, altro, la, verità, vie...
96    [organizza, dei, viaggi, nel, fine, settimana,...
97    [pesa, le, tue, valigie, prima, del, viaggio, ...
98    [è, molto, importante, non, perdere, di, vista...
99    [i, conigli, hanno, un, ottimo, udito, e, un, ...
Name: TEXT, Length: 100, dtype: object

<a id="Question 1 Answer English">
</a>

## Question-1 Answer - English

In [7]:
from sklearn.model_selection import train_test_split

# splits the training and test data set in 80% : 20%

english_train, english_test = train_test_split(dataframe_english['TEXT'], test_size=0.2, random_state=25)
#X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2, stratify=y)

print("\t English Language Train data: \n\n", english_train)
print('\n\n')
print("\t English Language Test data: \n\n", english_test)

	 English Language Train data: 

 80    [focus, on, quality, not, quantity, when, help...
25    [cut, out, the, cookies, decorate, them, and, ...
27    [shaded, areas, tend, to, be, soggier, and, so...
13    [if, you, publish, a, book, of, poetry, or, a,...
51    [when, you, finished, delete, the, history, on...
                            ...                        
61          [sugary, foods, can, irritate, the, throat]
15    [when, you, leave, the, seat, of, your, car, y...
90    [part, of, loving, life, is, to, not, be, rule...
62    [before, settlement, negotiations, you, should...
4     [look, for, a, partner, that, shares, your, le...
Name: TEXT, Length: 80, dtype: object



	 English Language Test data: 

 41    [try, doing, something, to, break, the, touch,...
29        [sleep, is, often, disturbed, by, depression]
33    [your, carbon, footprint, is, your, impact, on...
65    [you, ll, learn, how, to, play, by, interactin...
74    [many, people, see, their, hair, as, an, integ

In [8]:
# Displaying the shape of the Train and Test sets in English Language
print (english_train.shape)
print (english_test.shape)

(80,)
(20,)


<a id="Question 1 Answer Italian">
</a>


## Question-1 Answer - Italian

In [9]:
from sklearn.model_selection import train_test_split

# splits the training and test data set in 80% : 20%

italian_train, italian_test = train_test_split(dataframe_italian['TEXT'], test_size=0.2, random_state=25)
#X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2, stratify=y)

print("\t Italian Language Train data: \n\n", italian_train)
print('\n\n')
print("\t Italian Language Test data: \n\n", italian_test)

	 Italian Language Train data: 

 80    [le, erbacce, possono, soffocare, i, fagiolini...
25    [se, la, ricetta, ti, indica, di, cucinare, il...
27    [un, eccessivo, consumo, di, alcool, può, dann...
13    [osservare, un, campione, al, microscopio, per...
51    [se, la, tua, macchina, è, piccola, potresti, ...
                            ...                        
61    [non, è, semplice, essere, una, buona, moglie,...
15    [tenere, una, volpe, in, casa, è, illegale, in...
90    [le, migliori, sedie, da, ufficio, dovrebbero,...
62    [comprendere, il, motivo, per, cui, la, gente,...
4     [accendi, il, condizionatore, sull, aria, fred...
Name: TEXT, Length: 80, dtype: object



	 Italian Language Test data: 

 41    [quando, eri, più, giovane, probabilmente, ti,...
29    [se, lei, ti, piace, e, vuoi, che, smetta, di,...
33    [un, sari, è, un, abito, tradizionale, indossa...
65    [la, maschera, al, miele, ha, lo, scopo, di, n...
74    [connetti, l, altro, capo, del, cavo, alla, po

In [10]:
# Displaying the shape of the Train and Test sets in Italian Language
print (italian_train.shape)
print (italian_test.shape)

(80,)
(20,)


In [11]:
dataframe_english['TEXT']

0     [bring, up, academic, achievements, awards, an...
1     [please, list, people, you, have, helped, your...
2     [add, activated, carbon, straight, to, your, v...
3     [place, sensors, around, your, garden, and, wh...
4     [look, for, a, partner, that, shares, your, le...
                            ...                        
95    [rinse, your, face, with, warm, water, and, pa...
96    [staying, mentally, strong, means, winning, ha...
97    [the, person, who, has, the, highest, score, w...
98    [for, the, most, part, men, and, women, wear, ...
99    [look, at, the, woman, whom, you, are, listeni...
Name: TEXT, Length: 100, dtype: object

In [12]:
# Convert 100 lists of English words to Panda Series
list_to_series_for_english = pd.Series(dataframe_english['TEXT'])
english_list_to_series = list_to_series_for_english.apply(pd.Series).stack().reset_index(drop = True)
english_list_to_series

0              bring
1                 up
2           academic
3       achievements
4             awards
            ...     
1309             you
1310              or
1311             she
1312              is
1313        speaking
Length: 1314, dtype: object

In [13]:
# Convert 100 lists of Italian words to Panda Series
list_to_series_for_italian = pd.Series(dataframe_italian['TEXT'])
italian_list_to_series = list_to_series_for_italian.apply(pd.Series).stack().reset_index(drop = True)
italian_list_to_series

0          guardati
1                 i
2             piedi
3                 o
4               fai
           ...     
1301        possono
1302    individuare
1303              i
1304      predatori
1305     facilmente
Length: 1306, dtype: object

In [14]:
# Joining the Series in English as String
english_series_to_string = " ".join(english_list_to_series)
english_series_to_string

'bring up academic achievements awards and other milestones in your life please list people you have helped your personal achievements or troublesome times you have overcome add activated carbon straight to your vodka place sensors around your garden and when a cat comes in the motion activates a sensor look for a partner that shares your level of adventure in pursuing new types of experiences most animated action adventure thriller or mystery films and books could work here if people think you re phony or putting on airs they ll tell from a mile away place your tongue behind the hole in your lips and flick out bits of airs bake the fries until they are crispy and golden brown pour into a soufflé dish and bake until the top browns and puffs meditation can help your mind and body relax and can help you feel happier they are easier for the human body to digest making your overall experience more pleasant for example you could join a book club or take an art class if you publish a book of

In [15]:
# Joining the Series in Italian as String
italian_series_to_string = " ".join(italian_list_to_series)
italian_series_to_string

'guardati i piedi o fai finta di essere affascinata dalla copertina del tuo libro sei affascinato dal funzionamento della mente umana pensa ai tuoi sentimenti di amore passione e la sessualità quasi tutti sono in grado di ricevere amore e gentilezza accendi il condizionatore sull aria fredda ma non attivare la funzione di ricircolo chi vuole darsi delle arie di superiorità usa tono accondiscendente e battute dello stesso tipo osserva come lo zentangle si differenzia dalle altre forme d arte visiva tirare su il prezzo è una vera arte dunque inizia a coltivarla sin da ora attualmente la forma più comune di inquinamento acustico deriva dai trasporti soprattutto automobili motociclette e aeroplano guidare un automobile è una delle cose potenzialmente più pericolose che si possa fare adulti e bambini devono avere il passaporto per espatriare anche via mare o terra prima della data prevista della nascita del bambino devi stabilire chi sarà presente per aiutarti generalmente un campione di sa

<a id="Question 2 Answer-Ngram-model">
</a>

## Question-2 N-gram Function

In [16]:
# General Ngram Function
def ngram_function(ngram, input_string):
    ngra_eng = [input_string[i:i+ngram] for i in range(len(input_string)-ngram+1)]
    return ngra_eng

In [17]:
# Unigram Output for English
unigram_english = ngram_function(1, english_series_to_string)
unigram_english

['b',
 'r',
 'i',
 'n',
 'g',
 ' ',
 'u',
 'p',
 ' ',
 'a',
 'c',
 'a',
 'd',
 'e',
 'm',
 'i',
 'c',
 ' ',
 'a',
 'c',
 'h',
 'i',
 'e',
 'v',
 'e',
 'm',
 'e',
 'n',
 't',
 's',
 ' ',
 'a',
 'w',
 'a',
 'r',
 'd',
 's',
 ' ',
 'a',
 'n',
 'd',
 ' ',
 'o',
 't',
 'h',
 'e',
 'r',
 ' ',
 'm',
 'i',
 'l',
 'e',
 's',
 't',
 'o',
 'n',
 'e',
 's',
 ' ',
 'i',
 'n',
 ' ',
 'y',
 'o',
 'u',
 'r',
 ' ',
 'l',
 'i',
 'f',
 'e',
 ' ',
 'p',
 'l',
 'e',
 'a',
 's',
 'e',
 ' ',
 'l',
 'i',
 's',
 't',
 ' ',
 'p',
 'e',
 'o',
 'p',
 'l',
 'e',
 ' ',
 'y',
 'o',
 'u',
 ' ',
 'h',
 'a',
 'v',
 'e',
 ' ',
 'h',
 'e',
 'l',
 'p',
 'e',
 'd',
 ' ',
 'y',
 'o',
 'u',
 'r',
 ' ',
 'p',
 'e',
 'r',
 's',
 'o',
 'n',
 'a',
 'l',
 ' ',
 'a',
 'c',
 'h',
 'i',
 'e',
 'v',
 'e',
 'm',
 'e',
 'n',
 't',
 's',
 ' ',
 'o',
 'r',
 ' ',
 't',
 'r',
 'o',
 'u',
 'b',
 'l',
 'e',
 's',
 'o',
 'm',
 'e',
 ' ',
 't',
 'i',
 'm',
 'e',
 's',
 ' ',
 'y',
 'o',
 'u',
 ' ',
 'h',
 'a',
 'v',
 'e',
 ' ',
 'o',
 'v',
 'e'

In [18]:
# Length of Unigram English
len(unigram_english)

7189

In [19]:
# Bigram Output for English
bigram_english = ngram_function(2, english_series_to_string)
bigram_english

['br',
 'ri',
 'in',
 'ng',
 'g ',
 ' u',
 'up',
 'p ',
 ' a',
 'ac',
 'ca',
 'ad',
 'de',
 'em',
 'mi',
 'ic',
 'c ',
 ' a',
 'ac',
 'ch',
 'hi',
 'ie',
 'ev',
 've',
 'em',
 'me',
 'en',
 'nt',
 'ts',
 's ',
 ' a',
 'aw',
 'wa',
 'ar',
 'rd',
 'ds',
 's ',
 ' a',
 'an',
 'nd',
 'd ',
 ' o',
 'ot',
 'th',
 'he',
 'er',
 'r ',
 ' m',
 'mi',
 'il',
 'le',
 'es',
 'st',
 'to',
 'on',
 'ne',
 'es',
 's ',
 ' i',
 'in',
 'n ',
 ' y',
 'yo',
 'ou',
 'ur',
 'r ',
 ' l',
 'li',
 'if',
 'fe',
 'e ',
 ' p',
 'pl',
 'le',
 'ea',
 'as',
 'se',
 'e ',
 ' l',
 'li',
 'is',
 'st',
 't ',
 ' p',
 'pe',
 'eo',
 'op',
 'pl',
 'le',
 'e ',
 ' y',
 'yo',
 'ou',
 'u ',
 ' h',
 'ha',
 'av',
 've',
 'e ',
 ' h',
 'he',
 'el',
 'lp',
 'pe',
 'ed',
 'd ',
 ' y',
 'yo',
 'ou',
 'ur',
 'r ',
 ' p',
 'pe',
 'er',
 'rs',
 'so',
 'on',
 'na',
 'al',
 'l ',
 ' a',
 'ac',
 'ch',
 'hi',
 'ie',
 'ev',
 've',
 'em',
 'me',
 'en',
 'nt',
 'ts',
 's ',
 ' o',
 'or',
 'r ',
 ' t',
 'tr',
 'ro',
 'ou',
 'ub',
 'bl',
 'le',

In [20]:
len(bigram_english)

7188

In [21]:
# Trigram Output for English
trigram_english = ngram_function(3, english_series_to_string)
trigram_english

['bri',
 'rin',
 'ing',
 'ng ',
 'g u',
 ' up',
 'up ',
 'p a',
 ' ac',
 'aca',
 'cad',
 'ade',
 'dem',
 'emi',
 'mic',
 'ic ',
 'c a',
 ' ac',
 'ach',
 'chi',
 'hie',
 'iev',
 'eve',
 'vem',
 'eme',
 'men',
 'ent',
 'nts',
 'ts ',
 's a',
 ' aw',
 'awa',
 'war',
 'ard',
 'rds',
 'ds ',
 's a',
 ' an',
 'and',
 'nd ',
 'd o',
 ' ot',
 'oth',
 'the',
 'her',
 'er ',
 'r m',
 ' mi',
 'mil',
 'ile',
 'les',
 'est',
 'sto',
 'ton',
 'one',
 'nes',
 'es ',
 's i',
 ' in',
 'in ',
 'n y',
 ' yo',
 'you',
 'our',
 'ur ',
 'r l',
 ' li',
 'lif',
 'ife',
 'fe ',
 'e p',
 ' pl',
 'ple',
 'lea',
 'eas',
 'ase',
 'se ',
 'e l',
 ' li',
 'lis',
 'ist',
 'st ',
 't p',
 ' pe',
 'peo',
 'eop',
 'opl',
 'ple',
 'le ',
 'e y',
 ' yo',
 'you',
 'ou ',
 'u h',
 ' ha',
 'hav',
 'ave',
 've ',
 'e h',
 ' he',
 'hel',
 'elp',
 'lpe',
 'ped',
 'ed ',
 'd y',
 ' yo',
 'you',
 'our',
 'ur ',
 'r p',
 ' pe',
 'per',
 'ers',
 'rso',
 'son',
 'ona',
 'nal',
 'al ',
 'l a',
 ' ac',
 'ach',
 'chi',
 'hie',
 'iev',


In [22]:
len(trigram_english)

7187

In [23]:
# Unigram Output for Italian
unigram_italian = ngram_function(1, italian_series_to_string)
unigram_italian

['g',
 'u',
 'a',
 'r',
 'd',
 'a',
 't',
 'i',
 ' ',
 'i',
 ' ',
 'p',
 'i',
 'e',
 'd',
 'i',
 ' ',
 'o',
 ' ',
 'f',
 'a',
 'i',
 ' ',
 'f',
 'i',
 'n',
 't',
 'a',
 ' ',
 'd',
 'i',
 ' ',
 'e',
 's',
 's',
 'e',
 'r',
 'e',
 ' ',
 'a',
 'f',
 'f',
 'a',
 's',
 'c',
 'i',
 'n',
 'a',
 't',
 'a',
 ' ',
 'd',
 'a',
 'l',
 'l',
 'a',
 ' ',
 'c',
 'o',
 'p',
 'e',
 'r',
 't',
 'i',
 'n',
 'a',
 ' ',
 'd',
 'e',
 'l',
 ' ',
 't',
 'u',
 'o',
 ' ',
 'l',
 'i',
 'b',
 'r',
 'o',
 ' ',
 's',
 'e',
 'i',
 ' ',
 'a',
 'f',
 'f',
 'a',
 's',
 'c',
 'i',
 'n',
 'a',
 't',
 'o',
 ' ',
 'd',
 'a',
 'l',
 ' ',
 'f',
 'u',
 'n',
 'z',
 'i',
 'o',
 'n',
 'a',
 'm',
 'e',
 'n',
 't',
 'o',
 ' ',
 'd',
 'e',
 'l',
 'l',
 'a',
 ' ',
 'm',
 'e',
 'n',
 't',
 'e',
 ' ',
 'u',
 'm',
 'a',
 'n',
 'a',
 ' ',
 'p',
 'e',
 'n',
 's',
 'a',
 ' ',
 'a',
 'i',
 ' ',
 't',
 'u',
 'o',
 'i',
 ' ',
 's',
 'e',
 'n',
 't',
 'i',
 'm',
 'e',
 'n',
 't',
 'i',
 ' ',
 'd',
 'i',
 ' ',
 'a',
 'm',
 'o',
 'r',
 'e',
 ' '

In [24]:
# Length of Unigram Italian
len(unigram_italian)

7963

In [25]:
# Bigram Output for Italian
bigram_italian = ngram_function(2, italian_series_to_string)
bigram_italian

['gu',
 'ua',
 'ar',
 'rd',
 'da',
 'at',
 'ti',
 'i ',
 ' i',
 'i ',
 ' p',
 'pi',
 'ie',
 'ed',
 'di',
 'i ',
 ' o',
 'o ',
 ' f',
 'fa',
 'ai',
 'i ',
 ' f',
 'fi',
 'in',
 'nt',
 'ta',
 'a ',
 ' d',
 'di',
 'i ',
 ' e',
 'es',
 'ss',
 'se',
 'er',
 're',
 'e ',
 ' a',
 'af',
 'ff',
 'fa',
 'as',
 'sc',
 'ci',
 'in',
 'na',
 'at',
 'ta',
 'a ',
 ' d',
 'da',
 'al',
 'll',
 'la',
 'a ',
 ' c',
 'co',
 'op',
 'pe',
 'er',
 'rt',
 'ti',
 'in',
 'na',
 'a ',
 ' d',
 'de',
 'el',
 'l ',
 ' t',
 'tu',
 'uo',
 'o ',
 ' l',
 'li',
 'ib',
 'br',
 'ro',
 'o ',
 ' s',
 'se',
 'ei',
 'i ',
 ' a',
 'af',
 'ff',
 'fa',
 'as',
 'sc',
 'ci',
 'in',
 'na',
 'at',
 'to',
 'o ',
 ' d',
 'da',
 'al',
 'l ',
 ' f',
 'fu',
 'un',
 'nz',
 'zi',
 'io',
 'on',
 'na',
 'am',
 'me',
 'en',
 'nt',
 'to',
 'o ',
 ' d',
 'de',
 'el',
 'll',
 'la',
 'a ',
 ' m',
 'me',
 'en',
 'nt',
 'te',
 'e ',
 ' u',
 'um',
 'ma',
 'an',
 'na',
 'a ',
 ' p',
 'pe',
 'en',
 'ns',
 'sa',
 'a ',
 ' a',
 'ai',
 'i ',
 ' t',
 'tu',

In [26]:
len(bigram_italian)

7962

In [27]:
# Trigram Output for Italian
trigram_italian = ngram_function(3, italian_series_to_string)
trigram_italian

['gua',
 'uar',
 'ard',
 'rda',
 'dat',
 'ati',
 'ti ',
 'i i',
 ' i ',
 'i p',
 ' pi',
 'pie',
 'ied',
 'edi',
 'di ',
 'i o',
 ' o ',
 'o f',
 ' fa',
 'fai',
 'ai ',
 'i f',
 ' fi',
 'fin',
 'int',
 'nta',
 'ta ',
 'a d',
 ' di',
 'di ',
 'i e',
 ' es',
 'ess',
 'sse',
 'ser',
 'ere',
 're ',
 'e a',
 ' af',
 'aff',
 'ffa',
 'fas',
 'asc',
 'sci',
 'cin',
 'ina',
 'nat',
 'ata',
 'ta ',
 'a d',
 ' da',
 'dal',
 'all',
 'lla',
 'la ',
 'a c',
 ' co',
 'cop',
 'ope',
 'per',
 'ert',
 'rti',
 'tin',
 'ina',
 'na ',
 'a d',
 ' de',
 'del',
 'el ',
 'l t',
 ' tu',
 'tuo',
 'uo ',
 'o l',
 ' li',
 'lib',
 'ibr',
 'bro',
 'ro ',
 'o s',
 ' se',
 'sei',
 'ei ',
 'i a',
 ' af',
 'aff',
 'ffa',
 'fas',
 'asc',
 'sci',
 'cin',
 'ina',
 'nat',
 'ato',
 'to ',
 'o d',
 ' da',
 'dal',
 'al ',
 'l f',
 ' fu',
 'fun',
 'unz',
 'nzi',
 'zio',
 'ion',
 'ona',
 'nam',
 'ame',
 'men',
 'ent',
 'nto',
 'to ',
 'o d',
 ' de',
 'del',
 'ell',
 'lla',
 'la ',
 'a m',
 ' me',
 'men',
 'ent',
 'nte',
 'te ',


In [28]:
len(trigram_italian)

7961

In [29]:
# Counting the number of occurences of each character in Unigram English Train Set using FreqDist
from nltk.probability import FreqDist
english_unigram_fdist = FreqDist(unigram_english)
print(english_unigram_fdist)

# Counting the number of occurences of each character in Bigram English Train Set using FreqDist
english_bigram_fdist = FreqDist(bigram_english)
print(english_bigram_fdist)

# Counting the number of occurences of each character in Trigram English Train Set using FreqDist
english_trigram_fdist = FreqDist(trigram_english)
print(english_trigram_fdist)

<FreqDist with 30 samples and 7189 outcomes>
<FreqDist with 357 samples and 7188 outcomes>
<FreqDist with 1705 samples and 7187 outcomes>


In [30]:
english_unigram_fdist

FreqDist({' ': 1313, 'e': 686, 'o': 528, 't': 504, 'a': 470, 'i': 398, 'n': 398, 'r': 378, 's': 365, 'h': 273, ...})

In [31]:
# Storing the frequent character in Unigram English Train Set in a variable
english_unigram_fdist_common = english_unigram_fdist.most_common()
print("English Train Set Unigram \n")
print(english_unigram_fdist_common)
print("\n\n")

# Storing the frequent character in Bigram English Train Set in a variable
english_bigram_fdist_common = english_bigram_fdist.most_common()
print("English Train Set Bigram \n")
print(english_bigram_fdist_common)
print("\n\n")

# Storing the frequent character in Trigram English Train Set in a variable
english_trigram_fdist_common = english_trigram_fdist.most_common()
print("English Train Set Trigram \n")
print(english_trigram_fdist_common)

English Train Set Unigram 

[(' ', 1313), ('e', 686), ('o', 528), ('t', 504), ('a', 470), ('i', 398), ('n', 398), ('r', 378), ('s', 365), ('h', 273), ('l', 264), ('u', 207), ('d', 196), ('y', 180), ('c', 174), ('f', 148), ('p', 134), ('m', 134), ('g', 110), ('w', 89), ('b', 84), ('v', 72), ('k', 52), ('x', 10), ('z', 9), ('j', 6), ('q', 4), ('é', 1), ('3', 1), ('0', 1)]



English Train Set Bigram 

[('e ', 244), (' t', 187), (' a', 152), ('s ', 149), ('th', 136), ('he', 125), ('r ', 124), ('in', 121), ('ou', 118), ('n ', 110), ('t ', 110), ('d ', 101), ('an', 100), (' s', 86), (' o', 85), (' i', 82), (' y', 82), ('yo', 82), ('y ', 82), (' c', 81), ('er', 80), ('re', 75), ('es', 68), ('or', 66), (' p', 64), ('en', 63), ('ar', 62), (' w', 60), (' h', 59), ('te', 58), ('on', 57), (' f', 57), (' b', 56), ('o ', 55), ('ng', 54), ('to', 54), ('nd', 53), ('st', 53), ('ur', 53), ('at', 53), (' m', 52), ('g ', 49), ('l ', 49), ('ve', 48), ('nt', 45), ('u ', 45), ('f ', 45), (' l', 44), ('ea', 

In [32]:
# Counting the number of occurences of each character in Unigram Italian Train Set using FreqDist
italian_unigram_fdist = FreqDist(unigram_italian)
print(italian_unigram_fdist)

# Counting the number of occurences of each character in Bigram Italian Train Set using FreqDist
italian_bigram_fdist = FreqDist(bigram_italian)
print(italian_bigram_fdist)

# Counting the number of occurences of each character in Trigram Italian Train Set using FreqDist
italian_trigram_fdist = FreqDist(trigram_italian)
print(italian_trigram_fdist)

<FreqDist with 34 samples and 7963 outcomes>
<FreqDist with 289 samples and 7962 outcomes>
<FreqDist with 1476 samples and 7961 outcomes>


In [33]:
# Storing the frequent character in Unigram Italian Train Set in a variable
italian_unigram_fdist_common = italian_unigram_fdist.most_common()
print("Italian Train Set Unigram \n")
print(italian_unigram_fdist_common)
print("\n\n")

# Storing the frequent character in Bigram Italian Train Set in a variable
italian_bigram_fdist_common = italian_bigram_fdist.most_common()
print("Italian Train Set Bigram \n")
print(italian_bigram_fdist_common)
print("\n\n")

# Storing the frequent character in Trigram Italian Train Set in a variable
italian_trigram_fdist_common = italian_trigram_fdist.most_common()
print("Italian Train Set Trigram \n")
print(italian_trigram_fdist_common)

Italian Train Set Unigram 

[(' ', 1305), ('e', 785), ('i', 753), ('a', 713), ('o', 572), ('n', 481), ('r', 468), ('t', 467), ('l', 386), ('s', 314), ('c', 275), ('d', 243), ('p', 239), ('u', 237), ('m', 171), ('v', 110), ('g', 108), ('f', 91), ('b', 58), ('z', 55), ('h', 48), ('q', 29), ('è', 17), ('à', 13), ('ù', 10), ('ò', 4), ('0', 2), ('2', 2), ('5', 2), ('3', 1), ('k', 1), ('é', 1), ('ì', 1), ('1', 1)]



Italian Train Set Bigram 

[('e ', 336), ('i ', 272), ('a ', 242), ('o ', 224), (' p', 154), (' d', 152), ('re', 144), ('er', 124), (' c', 119), (' s', 113), (' a', 111), (' i', 101), ('ti', 96), ('on', 95), ('en', 95), ('di', 85), ('ar', 84), ('in', 83), (' l', 83), ('an', 83), ('n ', 80), ('ta', 78), ('co', 76), ('ri', 75), ('la', 73), ('te', 73), ('le', 73), ('nt', 71), ('de', 70), ('to', 69), ('al', 66), ('no', 66), ('or', 64), ('pe', 63), ('un', 63), ('ra', 63), (' e', 62), ('l ', 62), ('io', 62), ('st', 59), ('at', 57), (' m', 57), ('ne', 57), ('tt', 56), ('ll', 55), (' u'

In [34]:
# Creating a Dataframe for the frequent characters in Unigram English Train Set
english_train_set_uni_df = pd.DataFrame(english_unigram_fdist_common)
english_train_set_uni_df = english_train_set_uni_df.rename(columns={0: "English_Charcters", 1: "English_Frequency"})
english_train_set_uni_df

English_Charcters  English_Frequency
0                                  1313
1                  e                686
2                  o                528
3                  t                504
4                  a                470
5                  i                398
6                  n                398
7                  r                378
8                  s                365
9                  h                273
10                 l                264
11                 u                207
12                 d                196
13                 y                180
14                 c                174
15                 f                148
16                 p                134
17                 m                134
18                 g                110
19                 w                 89
20                 b                 84
21                 v                 72
22                 k                 52
23                 x                 10
24                 z                  9
25                 j                  6
26                 q                  4
27                 é                  1
28                 3                  1
29                 0                  1

In [35]:
# Creating a Dataframe for the frequent characters in Bigram English Train Set
english_train_set_bi_df = pd.DataFrame(english_bigram_fdist_common)
english_train_set_bi_df = english_train_set_bi_df.rename(columns={0: "English_Charcters", 1: "English_Frequency"})
english_train_set_bi_df

English_Charcters  English_Frequency
0                  e                 244
1                   t                187
2                   a                152
3                  s                 149
4                  th                136
..                ...                ...
352                hs                  1
353                nh                  1
354                ox                  1
355                iu                  1
356                sc                  1

[357 rows x 2 columns]

In [36]:
# Creating a Dataframe for the frequent characters in Trigram English Train Set
english_train_set_tri_df = pd.DataFrame(english_trigram_fdist_common)
english_train_set_tri_df = english_train_set_tri_df.rename(columns={0: "English_Charcters", 1: "English_Frequency"})
english_train_set_tri_df

English_Charcters  English_Frequency
0                   th                103
1                  the                 92
2                   yo                 81
3                  you                 81
4                  he                  66
...                ...                ...
1700               sam                  1
1701               hoe                  1
1702               oes                  1
1703               oma                  1
1704               pea                  1

[1705 rows x 2 columns]

In [37]:
# Creating a Dataframe for the frequent characters in Unigram Italian Train Set
italian_train_set_uni_df = pd.DataFrame(italian_unigram_fdist_common)
italian_train_set_uni_df = italian_train_set_uni_df.rename(columns={0: "Italian_Charcters", 1: "Italian_Frequency"})
italian_train_set_uni_df

Italian_Charcters  Italian_Frequency
0                                  1305
1                  e                785
2                  i                753
3                  a                713
4                  o                572
5                  n                481
6                  r                468
7                  t                467
8                  l                386
9                  s                314
10                 c                275
11                 d                243
12                 p                239
13                 u                237
14                 m                171
15                 v                110
16                 g                108
17                 f                 91
18                 b                 58
19                 z                 55
20                 h                 48
21                 q                 29
22                 è                 17
23                 à                 13
24                 ù                 10
25                 ò                  4
26                 0                  2
27                 2                  2
28                 5                  2
29                 3                  1
30                 k                  1
31                 é                  1
32                 ì                  1
33                 1                  1

In [38]:
# Creating a Dataframe for the frequent characters in Bigram Italian Train Set
italian_train_set_bi_df = pd.DataFrame(italian_bigram_fdist_common)
italian_train_set_bi_df = italian_train_set_bi_df.rename(columns={0: "Italian_Charcters", 1: "Italian_Frequency"})
italian_train_set_bi_df

Italian_Charcters  Italian_Frequency
0                  e                 336
1                  i                 272
2                  a                 242
3                  o                 224
4                   p                154
..                ...                ...
284                 2                  1
285                50                  1
286                oz                  1
287                oe                  1
288                sf                  1

[289 rows x 2 columns]

In [39]:
# Creating a Dataframe for the frequent characters in Trigram Italian Train Set
italian_train_set_tri_df = pd.DataFrame(italian_trigram_fdist_common)
italian_train_set_tri_df = italian_train_set_tri_df.rename(columns={0: "Italian_Charcters", 1: "Italian_Frequency"})
italian_train_set_tri_df

Italian_Charcters  Italian_Frequency
0                  re                  99
1                   di                 61
2                  di                  60
3                  la                  58
4                  ti                  56
...                ...                ...
1471                ud                  1
1472               dua                  1
1473               eda                  1
1474               aci                  1
1475               cil                  1

[1476 rows x 2 columns]

<a id="Question 2 Answer-Prob">
</a>

## Question-2 Probability

In [40]:
# Calculating the Probability for the most frequent words in English language
def probability_english(freq):
    prob =freq/len(english_series_to_string)
    return prob

In [41]:
english_train_set_uni_df['English_Probability'] = english_train_set_uni_df['English_Frequency'].apply(lambda freq: probability_english(freq))
english_train_set_uni_df

English_Charcters  English_Frequency  English_Probability
0                                  1313             0.182640
1                  e                686             0.095424
2                  o                528             0.073446
3                  t                504             0.070107
4                  a                470             0.065378
5                  i                398             0.055362
6                  n                398             0.055362
7                  r                378             0.052580
8                  s                365             0.050772
9                  h                273             0.037975
10                 l                264             0.036723
11                 u                207             0.028794
12                 d                196             0.027264
13                 y                180             0.025038
14                 c                174             0.024204
15                 f                148             0.020587
16                 p                134             0.018640
17                 m                134             0.018640
18                 g                110             0.015301
19                 w                 89             0.012380
20                 b                 84             0.011685
21                 v                 72             0.010015
22                 k                 52             0.007233
23                 x                 10             0.001391
24                 z                  9             0.001252
25                 j                  6             0.000835
26                 q                  4             0.000556
27                 é                  1             0.000139
28                 3                  1             0.000139
29                 0                  1             0.000139

In [42]:
english_train_set_bi_df['English_Probability'] = english_train_set_bi_df['English_Frequency'].apply(lambda freq: probability_english(freq))
english_train_set_bi_df

English_Charcters  English_Frequency  English_Probability
0                  e                 244             0.033941
1                   t                187             0.026012
2                   a                152             0.021143
3                  s                 149             0.020726
4                  th                136             0.018918
..                ...                ...                  ...
352                hs                  1             0.000139
353                nh                  1             0.000139
354                ox                  1             0.000139
355                iu                  1             0.000139
356                sc                  1             0.000139

[357 rows x 3 columns]

In [43]:
english_train_set_tri_df['English_Probability'] = english_train_set_tri_df['English_Frequency'].apply(lambda freq: probability_english(freq))
english_train_set_tri_df

English_Charcters  English_Frequency  English_Probability
0                   th                103             0.014327
1                  the                 92             0.012797
2                   yo                 81             0.011267
3                  you                 81             0.011267
4                  he                  66             0.009181
...                ...                ...                  ...
1700               sam                  1             0.000139
1701               hoe                  1             0.000139
1702               oes                  1             0.000139
1703               oma                  1             0.000139
1704               pea                  1             0.000139

[1705 rows x 3 columns]

In [44]:
# Calculating the Probability for the most frequent words in Italian language
def probability_italian(freq):
    prob =freq/len(italian_series_to_string)
    #prob =freq/len(unigram_string)
    return prob

In [45]:
italian_train_set_uni_df['Italian_Probability'] = italian_train_set_uni_df['Italian_Frequency'].apply(lambda freq: probability_italian(freq))
italian_train_set_uni_df

Italian_Charcters  Italian_Frequency  Italian_Probability
0                                  1305             0.163883
1                  e                785             0.098581
2                  i                753             0.094562
3                  a                713             0.089539
4                  o                572             0.071832
5                  n                481             0.060404
6                  r                468             0.058772
7                  t                467             0.058646
8                  l                386             0.048474
9                  s                314             0.039432
10                 c                275             0.034535
11                 d                243             0.030516
12                 p                239             0.030014
13                 u                237             0.029763
14                 m                171             0.021474
15                 v                110             0.013814
16                 g                108             0.013563
17                 f                 91             0.011428
18                 b                 58             0.007284
19                 z                 55             0.006907
20                 h                 48             0.006028
21                 q                 29             0.003642
22                 è                 17             0.002135
23                 à                 13             0.001633
24                 ù                 10             0.001256
25                 ò                  4             0.000502
26                 0                  2             0.000251
27                 2                  2             0.000251
28                 5                  2             0.000251
29                 3                  1             0.000126
30                 k                  1             0.000126
31                 é                  1             0.000126
32                 ì                  1             0.000126
33                 1                  1             0.000126

In [46]:
italian_train_set_bi_df['Italian_Probability'] = italian_train_set_bi_df['Italian_Frequency'].apply(lambda freq: probability_italian(freq))
italian_train_set_bi_df

Italian_Charcters  Italian_Frequency  Italian_Probability
0                  e                 336             0.042195
1                  i                 272             0.034158
2                  a                 242             0.030391
3                  o                 224             0.028130
4                   p                154             0.019339
..                ...                ...                  ...
284                 2                  1             0.000126
285                50                  1             0.000126
286                oz                  1             0.000126
287                oe                  1             0.000126
288                sf                  1             0.000126

[289 rows x 3 columns]

In [47]:
italian_train_set_tri_df['Italian_Probability'] = italian_train_set_tri_df['Italian_Frequency'].apply(lambda freq: probability_italian(freq))
italian_train_set_tri_df

Italian_Charcters  Italian_Frequency  Italian_Probability
0                  re                  99             0.012433
1                   di                 61             0.007660
2                  di                  60             0.007535
3                  la                  58             0.007284
4                  ti                  56             0.007033
...                ...                ...                  ...
1471                ud                  1             0.000126
1472               dua                  1             0.000126
1473               eda                  1             0.000126
1474               aci                  1             0.000126
1475               cil                  1             0.000126

[1476 rows x 3 columns]

In [48]:
both_lang_train_set_uni_df = pd.concat([english_train_set_uni_df, italian_train_set_uni_df], axis = 1, sort = True)
both_lang_train_set_uni_df

English_Charcters  English_Frequency  English_Probability  \
0                                1313.0             0.182640   
1                  e              686.0             0.095424   
2                  o              528.0             0.073446   
3                  t              504.0             0.070107   
4                  a              470.0             0.065378   
5                  i              398.0             0.055362   
6                  n              398.0             0.055362   
7                  r              378.0             0.052580   
8                  s              365.0             0.050772   
9                  h              273.0             0.037975   
10                 l              264.0             0.036723   
11                 u              207.0             0.028794   
12                 d              196.0             0.027264   
13                 y              180.0             0.025038   
14                 c              174.0             0.024204   
15                 f              148.0             0.020587   
16                 p              134.0             0.018640   
17                 m              134.0             0.018640   
18                 g              110.0             0.015301   
19                 w               89.0             0.012380   
20                 b               84.0             0.011685   
21                 v               72.0             0.010015   
22                 k               52.0             0.007233   
23                 x               10.0             0.001391   
24                 z                9.0             0.001252   
25                 j                6.0             0.000835   
26                 q                4.0             0.000556   
27                 é                1.0             0.000139   
28                 3                1.0             0.000139   
29                 0                1.0             0.000139   
30               NaN                NaN                  NaN   
31               NaN                NaN                  NaN   
32               NaN                NaN                  NaN   
33               NaN                NaN                  NaN   

   Italian_Charcters  Italian_Frequency  Italian_Probability  
0                                  1305             0.163883  
1                  e                785             0.098581  
2                  i                753             0.094562  
3                  a                713             0.089539  
4                  o                572             0.071832  
5                  n                481             0.060404  
6                  r                468             0.058772  
7                  t                467             0.058646  
8                  l                386             0.048474  
9                  s                314             0.039432  
10                 c                275             0.034535  
11                 d                243             0.030516  
12                 p                239             0.030014  
13                 u                237             0.029763  
14                 m                171             0.021474  
15                 v                110             0.013814  
16                 g                108             0.013563  
17                 f                 91             0.011428  
18                 b                 58             0.007284  
19                 z                 55             0.006907  
20                 h                 48             0.006028  
21                 q                 29             0.003642  
22                 è                 17             0.002135  
23                 à                 13             0.001633  
24                 ù                 10             0.001256  
25                 ò                  4             0.000502  
26                 0                  2             0.000251 

In [49]:
both_lang_train_set_bi_df = pd.concat([english_train_set_bi_df, italian_train_set_bi_df], axis = 1, sort = True)
both_lang_train_set_bi_df

English_Charcters  English_Frequency  English_Probability  \
0                  e                 244             0.033941   
1                   t                187             0.026012   
2                   a                152             0.021143   
3                  s                 149             0.020726   
4                  th                136             0.018918   
..                ...                ...                  ...   
352                hs                  1             0.000139   
353                nh                  1             0.000139   
354                ox                  1             0.000139   
355                iu                  1             0.000139   
356                sc                  1             0.000139   

    Italian_Charcters  Italian_Frequency  Italian_Probability  
0                  e               336.0             0.042195  
1                  i               272.0             0.034158  
2                  a               242.0             0.030391  
3                  o               224.0             0.028130  
4                   p              154.0             0.019339  
..                ...                ...                  ...  
352               NaN                NaN                  NaN  
353               NaN                NaN                  NaN  
354               NaN                NaN                  NaN  
355               NaN                NaN                  NaN  
356               NaN                NaN                  NaN  

[357 rows x 6 columns]

In [50]:
both_lang_train_set_tri_df = pd.concat([english_train_set_tri_df, italian_train_set_tri_df], axis = 1, sort = True)
both_lang_train_set_tri_df

English_Charcters  English_Frequency  English_Probability  \
0                   th                103             0.014327   
1                  the                 92             0.012797   
2                   yo                 81             0.011267   
3                  you                 81             0.011267   
4                  he                  66             0.009181   
...                ...                ...                  ...   
1700               sam                  1             0.000139   
1701               hoe                  1             0.000139   
1702               oes                  1             0.000139   
1703               oma                  1             0.000139   
1704               pea                  1             0.000139   

     Italian_Charcters  Italian_Frequency  Italian_Probability  
0                  re                99.0             0.012433  
1                   di               61.0             0.007660  
2                  di                60.0             0.007535  
3                  la                58.0             0.007284  
4                  ti                56.0             0.007033  
...                ...                ...                  ...  
1700               NaN                NaN                  NaN  
1701               NaN                NaN                  NaN  
1702               NaN                NaN                  NaN  
1703               NaN                NaN                  NaN  
1704               NaN                NaN                  NaN  

[1705 rows x 6 columns]

### Convert English Test set & Italian Test set to List of words

In [51]:
english_test

41    [try, doing, something, to, break, the, touch,...
29        [sleep, is, often, disturbed, by, depression]
33    [your, carbon, footprint, is, your, impact, on...
65    [you, ll, learn, how, to, play, by, interactin...
74    [many, people, see, their, hair, as, an, integ...
6     [if, people, think, you, re, phony, or, puttin...
57    [the, fertilizer, will, inspire, leafy, growth...
89    [leave, the, office, to, walk, across, the, st...
87    [ideally, this, drive, should, have, enough, s...
98    [for, the, most, part, men, and, women, wear, ...
0     [bring, up, academic, achievements, awards, an...
64    [it, s, important, that, you, learn, to, trust...
81    [the, file, type, you, pick, will, affect, the...
72    [each, state, has, an, office, of, voter, affa...
66    [dress, in, a, professional, manner, for, your...
91    [there, is, also, a, potential, to, suffocate,...
95    [rinse, your, face, with, warm, water, and, pa...
63    [if, you, do, not, have, a, valid, legal, 

In [52]:
# Convert 20 lists of English words to Panda Series
list_to_series_for_english_test_set = pd.Series(english_test)
english_test_set_list_to_series = list_to_series_for_english_test_set.apply(pd.Series).stack().reset_index(drop = True)
english_test_set_list_to_series

0            try
1          doing
2      something
3             to
4          break
         ...    
235          the
236         head
237           of
238         your
239          bed
Length: 240, dtype: object

In [53]:
# Convert Series of English words to single list
english_test_set_final_list = english_test_set_list_to_series.tolist()
english_test_set_final_list

['try',
 'doing',
 'something',
 'to',
 'break',
 'the',
 'touch',
 'jitters',
 'away',
 'by',
 'flirting',
 'physically',
 'sleep',
 'is',
 'often',
 'disturbed',
 'by',
 'depression',
 'your',
 'carbon',
 'footprint',
 'is',
 'your',
 'impact',
 'on',
 'the',
 'earth',
 's',
 'atmosphere',
 'you',
 'll',
 'learn',
 'how',
 'to',
 'play',
 'by',
 'interacting',
 'with',
 'those',
 'more',
 'talented',
 'than',
 'you',
 'many',
 'people',
 'see',
 'their',
 'hair',
 'as',
 'an',
 'integral',
 'part',
 'of',
 'their',
 'identity',
 'if',
 'people',
 'think',
 'you',
 're',
 'phony',
 'or',
 'putting',
 'on',
 'airs',
 'they',
 'll',
 'tell',
 'from',
 'a',
 'mile',
 'away',
 'the',
 'fertilizer',
 'will',
 'inspire',
 'leafy',
 'growth',
 'rather',
 'than',
 'flower',
 'growth',
 'leave',
 'the',
 'office',
 'to',
 'walk',
 'across',
 'the',
 'street',
 'to',
 'get',
 'coffee',
 'ideally',
 'this',
 'drive',
 'should',
 'have',
 'enough',
 'space',
 'to',
 'fit',
 'all',
 'of',
 'the',


In [54]:
italian_test

41    [quando, eri, più, giovane, probabilmente, ti,...
29    [se, lei, ti, piace, e, vuoi, che, smetta, di,...
33    [un, sari, è, un, abito, tradizionale, indossa...
65    [la, maschera, al, miele, ha, lo, scopo, di, n...
74    [connetti, l, altro, capo, del, cavo, alla, po...
6     [osserva, come, lo, zentangle, si, differenzia...
57    [riduci, i, carboidrati, raffinati, e, mangia,...
89       [tu, non, sei, i, pensieri, nella, tua, testa]
87        [percorri, strade, in, salita, con, la, bici]
98    [è, molto, importante, non, perdere, di, vista...
0     [guardati, i, piedi, o, fai, finta, di, essere...
64    [i, gattini, appena, nati, iniziano, a, nutrir...
81    [sappi, soltanto, che, sta, cercando, di, aiut...
72    [decidi, quante, persone, vuoi, che, facciano,...
66    [ci, sono, pettegolezzi, innocui, e, altri, ch...
91    [recati, presso, l, ufficio, del, tribunale, d...
95    [in, un, modo, o, nell, altro, la, verità, vie...
63    [quando, c, è, un, motivo, di, preoccupazi

In [55]:
# Convert 20 lists of English words to Panda Series
list_to_series_for_italian_test_set = pd.Series(italian_test)
italian_test_set_list_to_series = list_to_series_for_italian_test_set.apply(pd.Series).stack().reset_index(drop = True)
italian_test_set_list_to_series

0             quando
1                eri
2                più
3            giovane
4      probabilmente
           ...      
239                l
240     informazione
241              con
242              gli
243            amici
Length: 244, dtype: object

In [56]:
# Convert Series of Italian words to single list
italian_test_set_final_list = italian_test_set_list_to_series.tolist()
italian_test_set_final_list

['quando',
 'eri',
 'più',
 'giovane',
 'probabilmente',
 'ti',
 'piacevano',
 'i',
 'luoghi',
 'chiassosi',
 'e',
 'pieni',
 'di',
 'gente',
 'se',
 'lei',
 'ti',
 'piace',
 'e',
 'vuoi',
 'che',
 'smetta',
 'di',
 'detestarti',
 'devi',
 'scusarti',
 'un',
 'sari',
 'è',
 'un',
 'abito',
 'tradizionale',
 'indossato',
 'dalle',
 'donne',
 'indiane',
 'la',
 'maschera',
 'al',
 'miele',
 'ha',
 'lo',
 'scopo',
 'di',
 'nutrire',
 'e',
 'idratare',
 'la',
 'pelle',
 'e',
 'non',
 'di',
 'detergerla',
 'connetti',
 'l',
 'altro',
 'capo',
 'del',
 'cavo',
 'alla',
 'porta',
 'di',
 'ingresso',
 'del',
 'preamplificatore',
 'osserva',
 'come',
 'lo',
 'zentangle',
 'si',
 'differenzia',
 'dalle',
 'altre',
 'forme',
 'd',
 'arte',
 'visiva',
 'riduci',
 'i',
 'carboidrati',
 'raffinati',
 'e',
 'mangia',
 'più',
 'proteine',
 'magre',
 'e',
 'verdure',
 'tu',
 'non',
 'sei',
 'i',
 'pensieri',
 'nella',
 'tua',
 'testa',
 'percorri',
 'strade',
 'in',
 'salita',
 'con',
 'la',
 'bici',
 

### Converting English & Italian dataframe columns to dictionary containing each language Characters & Probabilities

In [57]:
# Converting two Unigram English columns from dataframe to dictionary
english_final_uni_dict = both_lang_train_set_uni_df.set_index('English_Charcters')['English_Probability'].to_dict()
english_final_uni_dict

{' ': 0.18264014466546113,
 'e': 0.09542356377799416,
 'o': 0.07344554179997217,
 't': 0.07010710808179163,
 'a': 0.06537766031436917,
 'i': 0.05536235915982751,
 'n': 0.05536235915982751,
 'r': 0.05258033106134372,
 's': 0.050772012797329256,
 'h': 0.0379746835443038,
 'l': 0.03672277089998609,
 'u': 0.028793990819307273,
 'd': 0.027263875365141188,
 'y': 0.025038252886354154,
 'c': 0.024203644456809014,
 'f': 0.02058700792878008,
 'p': 0.018639588259841425,
 'm': 0.018639588259841425,
 'g': 0.015301154541660872,
 'w': 0.012380025038252886,
 'b': 0.011684518013631937,
 'v': 0.010015301154541661,
 'k': 0.007233273056057866,
 'x': 0.0013910140492418973,
 'z': 0.0012519126443177077,
 'j': 0.0008346084295451384,
 'q': 0.0005564056196967589,
 'é': 0.00013910140492418973,
 '3': 0.00013910140492418973,
 '0': 0.00013910140492418973,
 nan: nan}

In [58]:
# Converting two Bigram English columns from dataframe to dictionary
english_final_bi_dict = both_lang_train_set_bi_df.set_index('English_Charcters')['English_Probability'].to_dict()
english_final_bi_dict

{'e ': 0.0339407428015023,
 ' t': 0.02601196272082348,
 ' a': 0.02114341354847684,
 's ': 0.020726109333704272,
 'th': 0.018917791069689802,
 'he': 0.017387675615523717,
 'r ': 0.017248574210599527,
 'in': 0.01683126999582696,
 'ou': 0.016413965781054387,
 'n ': 0.015301154541660872,
 't ': 0.015301154541660872,
 'd ': 0.014049241897343164,
 'an': 0.013910140492418973,
 ' s': 0.011962720823480317,
 ' o': 0.011823619418556128,
 ' i': 0.011406315203783558,
 ' y': 0.011406315203783558,
 'yo': 0.011406315203783558,
 'y ': 0.011406315203783558,
 ' c': 0.01126721379885937,
 'er': 0.011128112393935179,
 're': 0.01043260536931423,
 'es': 0.009458895534844901,
 'or': 0.009180692724996522,
 ' p': 0.008902489915148143,
 'en': 0.008763388510223954,
 'ar': 0.008624287105299763,
 ' w': 0.008346084295451384,
 ' h': 0.008206982890527193,
 'te': 0.008067881485603005,
 'on': 0.007928780080678814,
 ' f': 0.007928780080678814,
 ' b': 0.007789678675754625,
 'o ': 0.007650577270830436,
 'ng': 0.007511475865

In [59]:
# Converting two Trigram English columns from dataframe to dictionary
english_final_tri_dict = both_lang_train_set_tri_df.set_index('English_Charcters')['English_Probability'].to_dict()
english_final_tri_dict

{' th': 0.014327444707191543,
 'the': 0.012797329253025456,
 ' yo': 0.01126721379885937,
 'you': 0.01126721379885937,
 'he ': 0.009180692724996522,
 'nd ': 0.006955070246209487,
 ' an': 0.006815968841285297,
 'ng ': 0.006537766031436918,
 'ing': 0.006398664626512728,
 'ou ': 0.006259563221588538,
 ' to': 0.005981360411740158,
 're ': 0.005703157601891779,
 'es ': 0.005564056196967589,
 ' in': 0.005564056196967589,
 ' a ': 0.0054249547920434,
 'and': 0.00528585338711921,
 'our': 0.00528585338711921,
 'or ': 0.00528585338711921,
 'to ': 0.00528585338711921,
 'er ': 0.00514675198219502,
 'ur ': 0.005007650577270831,
 ' of': 0.004868549172346641,
 'e t': 0.004729447767422451,
 's t': 0.004312143552649882,
 'of ': 0.004033940742801502,
 ' ca': 0.0038948393378773127,
 ' co': 0.0038948393378773127,
 'on ': 0.0037557379329531226,
 'an ': 0.0037557379329531226,
 'in ': 0.0033384337181805537,
 'e a': 0.0033384337181805537,
 'll ': 0.0033384337181805537,
 'ed ': 0.003199332313256364,
 ' ha': 0.00

In [60]:
# Converting two Unigram Italian columns from dataframe to dictionary
italian_final_uni_dict = both_lang_train_set_uni_df.set_index('Italian_Charcters')['Italian_Probability'].to_dict()
italian_final_uni_dict

{' ': 0.1638829586839131,
 'e': 0.09858093683285193,
 'i': 0.09456235087278664,
 'a': 0.08953911842270501,
 'o': 0.07183222403616728,
 'n': 0.06040437021223157,
 'r': 0.058771819665955044,
 't': 0.058646238854703,
 'l': 0.04847419314328771,
 's': 0.039432374733140776,
 'c': 0.03453472309431119,
 'd': 0.030516137134245887,
 'p': 0.030013813889237725,
 'u': 0.029762652266733643,
 'm': 0.021474318724098957,
 'v': 0.013813889237724475,
 'g': 0.013562727615220394,
 'f': 0.011427853823935703,
 'b': 0.00728368705261836,
 'z': 0.006906944618862237,
 'h': 0.006027878940097953,
 'q': 0.00364184352630918,
 'è': 0.0021348737912846917,
 'à': 0.001632550546276529,
 'ù': 0.0012558081125204068,
 'ò': 0.0005023232450081628,
 '0': 0.0002511616225040814,
 '2': 0.0002511616225040814,
 '5': 0.0002511616225040814,
 '3': 0.0001255808112520407,
 'k': 0.0001255808112520407,
 'é': 0.0001255808112520407,
 'ì': 0.0001255808112520407,
 '1': 0.0001255808112520407}

In [61]:
# Converting two Bigram Italian columns from dataframe to dictionary
italian_final_bi_dict = both_lang_train_set_bi_df.set_index('Italian_Charcters')['Italian_Probability'].to_dict()
italian_final_bi_dict

{'e ': 0.04219515258068567,
 'i ': 0.034157980660555066,
 'a ': 0.030390556322993846,
 'o ': 0.028130101720457115,
 ' p': 0.019339444932814268,
 ' d': 0.019088283310310185,
 're': 0.01808363682029386,
 'er': 0.015572020595253045,
 ' c': 0.014944116538992842,
 ' s': 0.014190631671480597,
 ' a': 0.013939470048976516,
 ' i': 0.01268366193645611,
 'ti': 0.012055757880195906,
 'on': 0.011930177068943865,
 'en': 0.011930177068943865,
 'di': 0.010674368956423458,
 'ar': 0.010548788145171418,
 'in': 0.010423207333919377,
 ' l': 0.010423207333919377,
 'an': 0.010423207333919377,
 'n ': 0.010046464900163254,
 'ta': 0.009795303277659173,
 'co': 0.009544141655155092,
 'ri': 0.009418560843903051,
 'la': 0.00916739922139897,
 'te': 0.00916739922139897,
 'le': 0.00916739922139897,
 'nt': 0.008916237598894889,
 'de': 0.008790656787642848,
 'to': 0.008665075976390808,
 'al': 0.008288333542634686,
 'no': 0.008288333542634686,
 'or': 0.008037171920130605,
 'pe': 0.007911591108878563,
 'un': 0.00791159110

In [62]:
# Converting two Trigram Italian columns from dataframe to dictionary
italian_final_tri_dict = both_lang_train_set_tri_df.set_index('Italian_Charcters')['Italian_Probability'].to_dict()
italian_final_tri_dict

{'re ': 0.012432500313952028,
 ' di': 0.007660429486374482,
 'di ': 0.007534848675122442,
 'la ': 0.00728368705261836,
 'ti ': 0.007032525430114279,
 ' co': 0.006781363807610197,
 'le ': 0.006530202185106116,
 ' in': 0.006404621373854075,
 'ent': 0.006279040562602034,
 ' de': 0.0061534597513499935,
 'to ': 0.0061534597513499935,
 ' pe': 0.006027878940097953,
 'e d': 0.0059022981288459125,
 'no ': 0.005651136506341831,
 ' un': 0.005651136506341831,
 'te ': 0.00552555569508979,
 'ta ': 0.005148813261333669,
 'are': 0.005148813261333669,
 'per': 0.005023232450081627,
 ' pr': 0.005023232450081627,
 'e p': 0.004897651638829587,
 'e l': 0.004520909205073465,
 ' la': 0.004520909205073465,
 'o d': 0.004395328393821424,
 'ne ': 0.004395328393821424,
 'i s': 0.004269747582569383,
 'ell': 0.004144166771317343,
 'a p': 0.004144166771317343,
 ' e ': 0.004144166771317343,
 'i d': 0.004018585960065302,
 'del': 0.0038930051488132614,
 'con': 0.0038930051488132614,
 'men': 0.003767424337561221,
 'e s':

In [63]:
len(english_test_set_final_list)

240

In [64]:
english_test_set_final_list

['try',
 'doing',
 'something',
 'to',
 'break',
 'the',
 'touch',
 'jitters',
 'away',
 'by',
 'flirting',
 'physically',
 'sleep',
 'is',
 'often',
 'disturbed',
 'by',
 'depression',
 'your',
 'carbon',
 'footprint',
 'is',
 'your',
 'impact',
 'on',
 'the',
 'earth',
 's',
 'atmosphere',
 'you',
 'll',
 'learn',
 'how',
 'to',
 'play',
 'by',
 'interacting',
 'with',
 'those',
 'more',
 'talented',
 'than',
 'you',
 'many',
 'people',
 'see',
 'their',
 'hair',
 'as',
 'an',
 'integral',
 'part',
 'of',
 'their',
 'identity',
 'if',
 'people',
 'think',
 'you',
 're',
 'phony',
 'or',
 'putting',
 'on',
 'airs',
 'they',
 'll',
 'tell',
 'from',
 'a',
 'mile',
 'away',
 'the',
 'fertilizer',
 'will',
 'inspire',
 'leafy',
 'growth',
 'rather',
 'than',
 'flower',
 'growth',
 'leave',
 'the',
 'office',
 'to',
 'walk',
 'across',
 'the',
 'street',
 'to',
 'get',
 'coffee',
 'ideally',
 'this',
 'drive',
 'should',
 'have',
 'enough',
 'space',
 'to',
 'fit',
 'all',
 'of',
 'the',


<a id="Question 2 Answer-model-implemetation">
</a>

## Question-2 N-gram model Implementation

In [65]:
# Function for Unigram English Test set

# Input is given as list
def Unigram_English(input_test_set):
    count_eng = 0                                # Accuracy for English words
    count_ita = 0                                # Accuracy for Italian words
#     total_english_prob = 0
#     total_italian_prob = 0
    # Iterating through each item in the English test set
    for every_item in input_test_set:                  # 'every_item' have individual words from the list-Bring
        total_english_prob = 0
        total_italian_prob = 0
        char_ngram = ngram_function(1, every_item)     # 'char_ngram' have individual characters in the word obtained from prev step-['B', 'r', 'i', 'n', 'g'] 
        for char_index in range(0, len(char_ngram)):   # 'char_index' have the index of the individual charcters of prev step
            # Iterating over English Unigram Frequent items
            for ele in english_unigram_fdist:          
                if ele == (char_ngram[char_index]):
                    if ele in english_final_uni_dict.keys():
                        total_english_prob = total_english_prob + english_final_uni_dict[ele]
                    else:
                        break
            for ele in italian_unigram_fdist:
                if ele == (char_ngram[char_index]):
                    if ele in italian_final_uni_dict.keys():
                        total_italian_prob = total_italian_prob + italian_final_uni_dict[ele]
                    else:
                        break
        print("Total English Probability for the word is: ", total_english_prob)
        print("Total Italian Probability for the word is: ", total_italian_prob)
        if total_english_prob > total_italian_prob:
            print (every_item.upper()," is an English word!")
            print("\n")
        else: 
            print (every_item.upper()," is an Italian word!")   
            print("\n")
#         count_eng = 0
#         count_ita = 0
        if total_english_prob > total_italian_prob:
            count_eng = count_eng + 1
        else: 
            count_ita = count_ita + 1
#         return count_eng, count_ita
        print("Total count of Predicted English words: ", count_eng)
        print("Total count of Predicted Italian words: ", count_ita)
        english_accuracy = (count_eng/len(english_test_set_final_list))*100
        italian_accuracy = (count_ita/len(italian_test_set_final_list))*100
    print("Accuracy for English language is: ", english_accuracy)    
    print("Accuracy for Italian language is: ", italian_accuracy)   

In [66]:
Unigram_English(['bring', 'person'])

Total English Probability for the word is:  0.19029072193629154
Total Italian Probability for the word is:  0.23458495541881202
BRING  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.34622339685630826
Total Italian Probability for the word is:  0.3590355393695843
PERSON  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  2
Accuracy for English language is:  0.0
Accuracy for Italian language is:  0.819672131147541


### Unigram model output for English 

In [67]:
Unigram_English(english_test_set_final_list)

Total English Probability for the word is:  0.14772569202948949
Total Italian Probability for the word is:  0.11741805852065804
TRY  is an English word!


Total count of Predicted English words:  1
Total count of Predicted Italian words:  0
Total English Probability for the word is:  0.22673529002642925
Total Italian Probability for the word is:  0.27087780987065174
DOING  is an Italian word!


Total count of Predicted English words:  1
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.47238837112254833
Total Italian Probability for the word is:  0.46452342082129844
SOMETHING  is an English word!


Total count of Predicted English words:  2
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.1435526498817638
Total Italian Probability for the word is:  0.13047846289087028
TO  is an English word!


Total count of Predicted English words:  3
Total count of Predicted Italian words:  1
Total English Probability for 

AS  is an Italian word!


Total count of Predicted English words:  23
Total count of Predicted Italian words:  26
Total English Probability for the word is:  0.12074001947419669
Total Italian Probability for the word is:  0.1499434886349366
AN  is an Italian word!


Total count of Predicted English words:  23
Total count of Predicted Italian words:  27
Total English Probability for the word is:  0.44623730699680064
Total Italian Probability for the word is:  0.5225417556197413
INTEGRAL  is an Italian word!


Total count of Predicted English words:  23
Total count of Predicted Italian words:  28
Total English Probability for the word is:  0.20670468771734596
Total Italian Probability for the word is:  0.23697099083260076
PART  is an Italian word!


Total count of Predicted English words:  23
Total count of Predicted Italian words:  29
Total English Probability for the word is:  0.09403254972875226
Total Italian Probability for the word is:  0.08326007786010298
OF  is an English word!





Total count of Predicted English words:  59
Total count of Predicted Italian words:  71
Total English Probability for the word is:  0.12546946724161914
Total Italian Probability for the word is:  0.15320858972748963
IT  is an Italian word!


Total count of Predicted English words:  59
Total count of Predicted Italian words:  72
Total English Probability for the word is:  0.050772012797329256
Total Italian Probability for the word is:  0.039432374733140776
S  is an English word!


Total count of Predicted English words:  60
Total count of Predicted Italian words:  72
Total English Probability for the word is:  0.4796216441786062
Total Italian Probability for the word is:  0.5438904935325882
IMPORTANT  is an Italian word!


Total count of Predicted English words:  60
Total count of Predicted Italian words:  73
Total English Probability for the word is:  0.24356656002225624
Total Italian Probability for the word is:  0.21285947507220895
THAT  is an English word!


Total count of Predicte

In [68]:
italian_test_set_final_list

['quando',
 'eri',
 'più',
 'giovane',
 'probabilmente',
 'ti',
 'piacevano',
 'i',
 'luoghi',
 'chiassosi',
 'e',
 'pieni',
 'di',
 'gente',
 'se',
 'lei',
 'ti',
 'piace',
 'e',
 'vuoi',
 'che',
 'smetta',
 'di',
 'detestarti',
 'devi',
 'scusarti',
 'un',
 'sari',
 'è',
 'un',
 'abito',
 'tradizionale',
 'indossato',
 'dalle',
 'donne',
 'indiane',
 'la',
 'maschera',
 'al',
 'miele',
 'ha',
 'lo',
 'scopo',
 'di',
 'nutrire',
 'e',
 'idratare',
 'la',
 'pelle',
 'e',
 'non',
 'di',
 'detergerla',
 'connetti',
 'l',
 'altro',
 'capo',
 'del',
 'cavo',
 'alla',
 'porta',
 'di',
 'ingresso',
 'del',
 'preamplificatore',
 'osserva',
 'come',
 'lo',
 'zentangle',
 'si',
 'differenzia',
 'dalle',
 'altre',
 'forme',
 'd',
 'arte',
 'visiva',
 'riduci',
 'i',
 'carboidrati',
 'raffinati',
 'e',
 'mangia',
 'più',
 'proteine',
 'magre',
 'e',
 'verdure',
 'tu',
 'non',
 'sei',
 'i',
 'pensieri',
 'nella',
 'tua',
 'testa',
 'percorri',
 'strade',
 'in',
 'salita',
 'con',
 'la',
 'bici',
 

In [69]:
# Function for Unigram Italian Test set

# Input is given as list
def Unigram_Italian(input_test_set):
    count_eng = 0                                # Accuracy for English words
    count_ita = 0                                # Accuracy for Italian words
#     total_english_prob = 0
#     total_italian_prob = 0
    # Iterating through each item in the Italian test set
    for every_item in input_test_set:                  # 'every_item' have individual words from the list-Bring
        total_english_prob = 0
        total_italian_prob = 0
        char_ngram = ngram_function(1, every_item)     # 'char_ngram' have individual characters in the word obtained from prev step-['B', 'r', 'i', 'n', 'g'] 
        for char_index in range(0, len(char_ngram)):   # 'char_index' have the index of the individual charcters of prev step
            # Iterating over Italian Unigram Frequent items
            for ele in english_unigram_fdist:          
                if ele == (char_ngram[char_index]):
                    if ele in english_final_uni_dict.keys():
                        total_english_prob = total_english_prob + english_final_uni_dict[ele]
                    else:
                        break
            for ele in italian_unigram_fdist:
                if ele == (char_ngram[char_index]):
                    if ele in italian_final_uni_dict.keys():
                        total_italian_prob = total_italian_prob + italian_final_uni_dict[ele]
                    else:
                        break
        print("Total English Probability for the word is: ", total_english_prob)
        print("Total Italian Probability for the word is: ", total_italian_prob)
        if total_english_prob > total_italian_prob:
            print (every_item.upper()," is an English word!")
            print("\n")
        else: 
            print (every_item.upper()," is an Italian word!")   
            print("\n")
#         count_eng = 0
#         count_ita = 0
        if total_english_prob > total_italian_prob:
            count_eng = count_eng + 1
        else: 
            count_ita = count_ita + 1
#         return count_eng, count_ita
        print("Total count of Predicted English words: ", count_eng)
        print("Total count of Predicted Italian words: ", count_ita)
        english_accuracy = (count_eng/len(english_test_set_final_list))*100
        italian_accuracy = (count_ita/len(italian_test_set_final_list))*100
    print("Accuracy for English language is: ", english_accuracy)    
    print("Accuracy for Italian language is: ", italian_accuracy)

### Unigram model output for Italian 

In [70]:
Unigram_Italian(italian_test_set_final_list)

Total English Probability for the word is:  0.2507998330783141
Total Italian Probability for the word is:  0.2856963455983926
QUANDO  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.2033662539991654
Total Italian Probability for the word is:  0.2519151073715936
ERI  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  2
Total English Probability for the word is:  0.07400194741966894
Total Italian Probability for the word is:  0.12583197287454478
PIÙ  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  3
Total English Probability for the word is:  0.37028793990819303
Total Italian Probability for the word is:  0.44229561722968724
GIOVANE  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  4
Total English Probability for the


Total count of Predicted English words:  20
Total count of Predicted Italian words:  186
Total English Probability for the word is:  0.03672277089998609
Total Italian Probability for the word is:  0.04847419314328771
L  is an Italian word!


Total count of Predicted English words:  20
Total count of Predicted Italian words:  187
Total English Probability for the word is:  0.29225205174572266
Total Italian Probability for the word is:  0.37347733266356903
ANSIA  is an Italian word!


Total count of Predicted English words:  20
Total count of Predicted Italian words:  188
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0.0021348737912846917
È  is an Italian word!


Total count of Predicted English words:  20
Total count of Predicted Italian words:  189
Total English Probability for the word is:  0.14953401029350397
Total Italian Probability for the word is:  0.17970614090167022
UNA  is an Italian word!


Total count of Predicted English words:  

<a id="Question 2 Answer-accuracy">
</a>

## Question-2 Unigram model Accuracy Analysis

In the Unigram model, the words in the test set is split into single charcters and calculated the probability. The accuracy of the Unigram model is: 

- Around 43% and 56% for English and Italian Language respectively for the English test set. 
- Around 9.5% and 90.5% for English and Italian Language respectively for the Italian test set. 

According to the above reported accuracies, the unigram model showed somewhat poor performance in English test set words since, 43% of the words are predicted wrongly. But on seeing the unigram model performance in Italian test set words, the model really performed well in predicting the characters, passing with 90.5%.
This shows that unigram model fails to prove as a great working model in predicting both the language words.

In [71]:
# Function for Bigram English Test set

# Input is given as list
def Bigram_English(input_test_set):
    count_eng = 0                                # Accuracy for English words
    count_ita = 0                                # Accuracy for Italian words
#     total_english_prob = 0
#     total_italian_prob = 0
    # Iterating through each item in the English test set
    for every_item in input_test_set:                  # 'every_item' have individual words from the list-Bring
        total_english_prob = 0
        total_italian_prob = 0
        char_ngram = ngram_function(2, every_item)     # 'char_ngram' have individual characters in the word obtained from prev step-['B', 'r', 'i', 'n', 'g'] 
        for char_index in range(0, len(char_ngram)):   # 'char_index' have the index of the individual charcters of prev step
            # Iterating over English Bigram Frequent items
            for ele in english_bigram_fdist:          
                if ele == (char_ngram[char_index]):
                    if ele in english_final_bi_dict.keys():
                        total_english_prob = total_english_prob + english_final_bi_dict[ele]
                    else:
                        break
            for ele in italian_bigram_fdist:
                if ele == (char_ngram[char_index]):
                    if ele in italian_final_bi_dict.keys():
                        total_italian_prob = total_italian_prob + italian_final_bi_dict[ele]
                    else:
                        break
        print("Total English Probability for the word is: ", total_english_prob)
        print("Total Italian Probability for the word is: ", total_italian_prob)
        if total_english_prob > total_italian_prob:
            print (every_item.upper()," is an English word!")
            print("\n")
        else: 
            print (every_item.upper()," is an Italian word!")   
            print("\n")
#         count_eng = 0
#         count_ita = 0
        if total_english_prob > total_italian_prob:
            count_eng = count_eng + 1
        else: 
            count_ita = count_ita + 1
#         return count_eng, count_ita
        print("Total count of Predicted English words: ", count_eng)
        print("Total count of Predicted Italian words: ", count_ita)
        english_accuracy = (count_eng/len(english_test_set_final_list))*100
        italian_accuracy = (count_ita/len(italian_test_set_final_list))*100
    print("Accuracy for English language is: ", english_accuracy)    
    print("Accuracy for Italian language is: ", italian_accuracy)

### Bigram model output for English

In [72]:
Bigram_English(english_test_set_final_list)

Total English Probability for the word is:  0.00486854917234664
Total Italian Probability for the word is:  0.00539997488383775
TRY  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.027263875365141188
Total Italian Probability for the word is:  0.01682782870777345
DOING  is an English word!


Total count of Predicted English words:  1
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.06301293643065795
Total Italian Probability for the word is:  0.030892879568002008
SOMETHING  is an English word!


Total count of Predicted English words:  2
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.007511475865906245
Total Italian Probability for the word is:  0.008665075976390808
TO  is an Italian word!


Total count of Predicted English words:  2
Total count of Predicted Italian words:  2
Total English Probability

Total English Probability for the word is:  0.013771039087494783
Total Italian Probability for the word is:  0.01117669220143162
SAVED  is an English word!


Total count of Predicted English words:  93
Total count of Predicted Italian words:  62
Total English Probability for the word is:  0.013771039087494783
Total Italian Probability for the word is:  0.008162752731382644
EACH  is an English word!


Total count of Predicted English words:  94
Total count of Predicted Italian words:  62
Total English Probability for the word is:  0.02670746974544443
Total Italian Probability for the word is:  0.03353007660429486
STATE  is an Italian word!


Total count of Predicted English words:  94
Total count of Predicted Italian words:  63
Total English Probability for the word is:  0.009041591320072333
Total Italian Probability for the word is:  0.004144166771317343
HAS  is an English word!


Total count of Predicted English words:  95
Total count of Predicted Italian words:  63
Total English Prob

In [73]:
# Function for Bigram Italian Test set

# Input is given as list
def Bigram_Italian(input_test_set):
    count_eng = 0                                # Accuracy for English words
    count_ita = 0                                # Accuracy for Italian words
#     total_english_prob = 0
#     total_italian_prob = 0
    # Iterating through each item in the Italian test set
    for every_item in input_test_set:                  # 'every_item' have individual words from the list-Bring
        total_english_prob = 0
        total_italian_prob = 0
        char_ngram = ngram_function(2, every_item)     # 'char_ngram' have individual characters in the word obtained from prev step-['B', 'r', 'i', 'n', 'g'] 
        for char_index in range(0, len(char_ngram)):   # 'char_index' have the index of the individual charcters of prev step
            # Iterating over Italian Bigram Frequent items
            for ele in english_bigram_fdist:          
                if ele == (char_ngram[char_index]):
                    if ele in english_final_bi_dict.keys():
                        total_english_prob = total_english_prob + english_final_bi_dict[ele]
                    else:
                        break
            for ele in italian_bigram_fdist:
                if ele == (char_ngram[char_index]):
                    if ele in italian_final_bi_dict.keys():
                        total_italian_prob = total_italian_prob + italian_final_bi_dict[ele]
                    else:
                        break
        print("Total English Probability for the word is: ", total_english_prob)
        print("Total Italian Probability for the word is: ", total_italian_prob)
        if total_english_prob > total_italian_prob:
            print (every_item.upper()," is an English word!")
            print("\n")
        else: 
            print (every_item.upper()," is an Italian word!")   
            print("\n")
#         count_eng = 0
#         count_ita = 0
        if total_english_prob > total_italian_prob:
            count_eng = count_eng + 1
        else: 
            count_ita = count_ita + 1
#         return count_eng, count_ita
        print("Total count of Predicted English words: ", count_eng)
        print("Total count of Predicted Italian words: ", count_ita)
        english_accuracy = (count_eng/len(english_test_set_final_list))*100
        italian_accuracy = (count_ita/len(italian_test_set_final_list))*100
    print("Accuracy for English language is: ", english_accuracy)    
    print("Accuracy for Italian language is: ", italian_accuracy)

### Bigram model output for Italian

In [74]:
Bigram_Italian(italian_test_set_final_list)

Total English Probability for the word is:  0.024481847266657395
Total Italian Probability for the word is:  0.027502197664196912
QUANDO  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.014744748921964111
Total Italian Probability for the word is:  0.024990581439156098
ERI  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  2
Total English Probability for the word is:  0.0016692168590902768
Total Italian Probability for the word is:  0.005274394072585709
PIÙ  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  3
Total English Probability for the word is:  0.02517735429127834
Total Italian Probability for the word is:  0.033781238226798946
GIOVANE  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  4
Total English Probabi

Total count of Predicted English words:  5
Total count of Predicted Italian words:  133
Total English Probability for the word is:  0.004033940742801502
Total Italian Probability for the word is:  0.010423207333919379
SAPPI  is an Italian word!


Total count of Predicted English words:  5
Total count of Predicted Italian words:  134
Total English Probability for the word is:  0.03741827792460704
Total Italian Probability for the word is:  0.050106743689564234
SOLTANTO  is an Italian word!


Total count of Predicted English words:  5
Total count of Predicted Italian words:  135
Total English Probability for the word is:  0.02100431214355265
Total Italian Probability for the word is:  0.007786010297626523
CHE  is an English word!


Total count of Predicted English words:  6
Total count of Predicted Italian words:  135
Total English Probability for the word is:  0.01126721379885937
Total Italian Probability for the word is:  0.017204571141529575
STA  is an Italian word!


Total count of P

<a id="Question 3">
</a>

## Question-3  Bigram model Accuracy Analysis

In the Bigram model, the words in the test set is split into double(two) charcters and calculated the probability. The accuracy of the Bigram model is: 

- Around 57% and 42% for English and Italian Language respectively for the English test set. 
- Around 5% and 94.6% for English and Italian Language respectively for the Italian test set. 

According to the above reported accuracies, the bigram model showed improved performance in predicting the English test set words when compared to the unigram model English words accuracy. On considering the bigram model performance in Italian test set words, the model really performed absolutely well in predicting the characters, passing with 94.6%.
This shows that the bigram model was better when compared to the unigram model since bigram model has very good performance in predicting both the language words.

In [75]:
# Function for Trigram English Test set

# Input is given as list
def Trigram_English(input_test_set):
    count_eng = 0                                # Accuracy for English words
    count_ita = 0                                # Accuracy for Italian words
#     total_english_prob = 0
#     total_italian_prob = 0
    # Iterating through each item in the English test set
    for every_item in input_test_set:                  # 'every_item' have individual words from the list-Bring
        total_english_prob = 0
        total_italian_prob = 0
        char_ngram = ngram_function(3, every_item)     # 'char_ngram' have individual characters in the word obtained from prev step-['B', 'r', 'i', 'n', 'g'] 
        for char_index in range(0, len(char_ngram)):   # 'char_index' have the index of the individual charcters of prev step
            # Iterating over English Trigram Frequent items
            for ele in english_trigram_fdist:          
                if ele == (char_ngram[char_index]):
                    if ele in english_final_tri_dict.keys():
                        total_english_prob = total_english_prob + english_final_tri_dict[ele]
                    else:
                        break
            for ele in italian_trigram_fdist:
                if ele == (char_ngram[char_index]):
                    if ele in italian_final_tri_dict.keys():
                        total_italian_prob = total_italian_prob + italian_final_tri_dict[ele]
                    else:
                        break
        print("Total English Probability for the word is: ", total_english_prob)
        print("Total Italian Probability for the word is: ", total_italian_prob)
        if total_english_prob > total_italian_prob:
            print (every_item.upper()," is an English word!")
            print("\n")
        else: 
            print (every_item.upper()," is an Italian word!")   
            print("\n")
#         count_eng = 0
#         count_ita = 0
        if total_english_prob > total_italian_prob:
            count_eng = count_eng + 1
        else: 
            count_ita = count_ita + 1
#         return count_eng, count_ita
        print("Total count of Predicted English words: ", count_eng)
        print("Total count of Predicted Italian words: ", count_ita)
        english_accuracy = (count_eng/len(english_test_set_final_list))*100
        italian_accuracy = (count_ita/len(italian_test_set_final_list))*100
    print("Accuracy for English language is: ", english_accuracy)    
    print("Accuracy for Italian language is: ", italian_accuracy)

### Trigram model output for English

In [76]:
Trigram_English(english_test_set_final_list)

Total English Probability for the word is:  0.0005564056196967589
Total Italian Probability for the word is:  0
TRY  is an English word!


Total count of Predicted English words:  1
Total count of Predicted Italian words:  0
Total English Probability for the word is:  0.006815968841285297
Total Italian Probability for the word is:  0.0002511616225040814
DOING  is an English word!


Total count of Predicted English words:  2
Total count of Predicted Italian words:  0
Total English Probability for the word is:  0.011545416608707747
Total Italian Probability for the word is:  0.0016325505462765292
SOMETHING  is an English word!


Total count of Predicted English words:  3
Total count of Predicted Italian words:  0
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0
TO  is an Italian word!


Total count of Predicted English words:  3
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.0018083182640144665
Total Ita


Total count of Predicted English words:  82
Total count of Predicted Italian words:  42
Total English Probability for the word is:  0.00528585338711921
Total Italian Probability for the word is:  0.0022604546025367326
AND  is an English word!


Total count of Predicted English words:  83
Total count of Predicted Italian words:  42
Total English Probability for the word is:  0.016413965781054387
Total Italian Probability for the word is:  0.0001255808112520407
OTHER  is an English word!


Total count of Predicted English words:  84
Total count of Predicted Italian words:  42
Total English Probability for the word is:  0.006259563221588538
Total Italian Probability for the word is:  0.007032525430114279
MILESTONES  is an Italian word!


Total count of Predicted English words:  84
Total count of Predicted Italian words:  43
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0
IN  is an Italian word!


Total count of Predicted English words:  84
Tota

LEGAL  is an English word!


Total count of Predicted English words:  131
Total count of Predicted Italian words:  70
Total English Probability for the word is:  0.0005564056196967589
Total Italian Probability for the word is:  0.0002511616225040814
CLAIM  is an English word!


Total count of Predicted English words:  132
Total count of Predicted Italian words:  70
Total English Probability for the word is:  0.0018083182640144665
Total Italian Probability for the word is:  0
ANY  is an English word!


Total count of Predicted English words:  133
Total count of Predicted Italian words:  70
Total English Probability for the word is:  0.001530115454166087
Total Italian Probability for the word is:  0.0001255808112520407
LAWSUIT  is an English word!


Total count of Predicted English words:  134
Total count of Predicted Italian words:  70
Total English Probability for the word is:  0.01126721379885937
Total Italian Probability for the word is:  0
YOU  is an English word!


Total count of P

In [77]:
# Function for Trigram Italian Test set

# Input is given as list
def Trigram_Italian(input_test_set):
    count_eng = 0                                # Accuracy for English words
    count_ita = 0                                # Accuracy for Italian words
#     total_english_prob = 0
#     total_italian_prob = 0
    # Iterating through each item in the Italian test set
    for every_item in input_test_set:                  # 'every_item' have individual words from the list-Bring
        total_english_prob = 0
        total_italian_prob = 0
        char_ngram = ngram_function(3, every_item)     # 'char_ngram' have individual characters in the word obtained from prev step-['B', 'r', 'i', 'n', 'g'] 
        for char_index in range(0, len(char_ngram)):   # 'char_index' have the index of the individual charcters of prev step
            # Iterating over Italian Bigram Frequent items
            for ele in english_trigram_fdist:          
                if ele == (char_ngram[char_index]):
                    if ele in english_final_tri_dict.keys():
                        total_english_prob = total_english_prob + english_final_tri_dict[ele]
                    else:
                        break
            for ele in italian_trigram_fdist:
                if ele == (char_ngram[char_index]):
                    if ele in italian_final_tri_dict.keys():
                        total_italian_prob = total_italian_prob + italian_final_tri_dict[ele]
                    else:
                        break
        print("Total English Probability for the word is: ", total_english_prob)
        print("Total Italian Probability for the word is: ", total_italian_prob)
        if total_english_prob > total_italian_prob:
#             list_eng = []
#             list_eng.extend(every_item)
#             print(list_eng)
            print (every_item.upper()," is an English word!")
            print("\n")
        else: 
#             list_ita = []
#             list_ita.extend(every_item)
#             print(list_ita)
            print (every_item.upper()," is an Italian word!")   
            print("\n")
#         count_eng = 0
#         count_ita = 0
        if total_english_prob > total_italian_prob:
            count_eng = count_eng + 1
        else: 
            count_ita = count_ita + 1
#         return count_eng, count_ita
        print("Total count of Predicted English words: ", count_eng)
        print("Total count of Predicted Italian words: ", count_ita)
        english_accuracy = (count_eng/len(english_test_set_final_list))*100
        italian_accuracy = (count_ita/len(italian_test_set_final_list))*100
    print("Accuracy for English language is: ", english_accuracy)    
    print("Accuracy for Italian language is: ", italian_accuracy)

### Trigram model output for Italian

In [78]:
Trigram_Italian(italian_test_set_final_list)

Total English Probability for the word is:  0.005842259006815969
Total Italian Probability for the word is:  0.008037171920130605
QUANDO  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  1
Total English Probability for the word is:  0.0004173042147725692
Total Italian Probability for the word is:  0.0017581313575285697
ERI  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  2
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0.0012558081125204068
PIÙ  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  3
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0.0036418435263091795
GIOVANE  is an Italian word!


Total count of Predicted English words:  0
Total count of Predicted Italian words:  4
Total English Probability for the word is:  0.006259563

Total English Probability for the word is:  0.0008346084295451384
Total Italian Probability for the word is:  0.0018837121687806102
APPENA  is an Italian word!


Total count of Predicted English words:  8
Total count of Predicted Italian words:  122
Total English Probability for the word is:  0.0011128112393935178
Total Italian Probability for the word is:  0.002888358658796936
NATI  is an Italian word!


Total count of Predicted English words:  8
Total count of Predicted Italian words:  123
Total English Probability for the word is:  0.0004173042147725692
Total Italian Probability for the word is:  0.006906944618862238
INIZIANO  is an Italian word!


Total count of Predicted English words:  8
Total count of Predicted Italian words:  124
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0
A  is an Italian word!


Total count of Predicted English words:  8
Total count of Predicted Italian words:  125
Total English Probability for the word is:  0.0



Total count of Predicted English words:  12
Total count of Predicted Italian words:  204
Total English Probability for the word is:  0.002086521073862846
Total Italian Probability for the word is:  0.005902298128845913
IMMAGINARIO  is an Italian word!


Total count of Predicted English words:  12
Total count of Predicted Italian words:  205
Total English Probability for the word is:  0.0006955070246209487
Total Italian Probability for the word is:  0.004144166771317343
ORGANIZZA  is an Italian word!


Total count of Predicted English words:  12
Total count of Predicted Italian words:  206
Total English Probability for the word is:  0
Total Italian Probability for the word is:  0
UN  is an Italian word!


Total count of Predicted English words:  12
Total count of Predicted Italian words:  207
Total English Probability for the word is:  0.0011128112393935178
Total Italian Probability for the word is:  0.004018585960065302
PROGRAMMA  is an Italian word!


Total count of Predicted Englis

<a id="Question 4">
</a>

## Question-4  Trigram model Accuracy Analysis

In the Trigram model, the words in the test set is split into triple(three) charcters and calculated the probability. The accuracy of the Trigram model is: 

- Around 65% and 35% for English and Italian Language respectively for the English test set. 
- Around 5% and 95% for English and Italian Language respectively for the Italian test set. 

According to the above reported accuracies, the trigram model performed better than bigram model in predicting the English test set words with 7% increase in predicting the English words correctly as English. On considering the trigram model performance in Italian test set words, the model outperformed both unigram and bigram models. The trigram model showed an accuracy of 95% in predicting the Italian test set words which proves to be a better predicting model.
This shows that the trigram model is slightly better than bigram model and far better than unigram model in predicting both the language words.